### ISSUE TITLE
Flatpak doesn't follow all dependencies across repos

### ISSUE THREAD

**Issue by nanonyme** [flatpak/flatpak#3094]
This has been a known and accepted issue but there's now a real use-case with Flathub beta that's breaking because of this. When I install application X from Flathub beta and I have both Flathub beta and Flathub installed as repositories, I would expect all dependency runtimes and extensions to be installed from Flathub if they are not available in Flathub beta. At least app extensions seem not to be correctly picked up from other repos (which was noticed with Steam Flatpak).

**Comment by mwleeds** [comment]
It would be helpful if you gave steps to reproduce, including steps to add the beta remote.

**Comment by gasinvein** [comment]
This is also reproducible with Lutris (from Flathub Beta), flatpak doesn't auto install app extensions `.Compat.i386` and `.GL32` (from Flathub): https://github.com/flathub/net.lutris.Lutris/issues/53#issuecomment-531472149

**Comment by nanonyme** [comment]
Right, so I guess this is in fact a duplicate of https://github.com/flatpak/flatpak/issues/3072

**Comment by gasinvein** [comment]
@nanonyme Yeah, looks like it. Yet this one is more generic.

**Comment by nanonyme** [comment]
Well, the generic issue is not new. Currently basically every producer of Flatpak stuff (eg Gnome master) must mirror everything their upstream producers make to avoid dependency issues. Flathub beta just further promotes this matter as agreement with @alexlarsson was runtime is only published (along with its extensions) to either beta or main but not both so things remain unambiguous.

**Comment by A6GibKm** [comment]
@alexlarsson, do you think it could be possible to mark certain repos as 'children/parent' of another repo, so dependencies can be followed in at least one way? There are many apps running into this issue at `flathub-beta`, which should at least trust the contents of `flathub`.

The issues I have seen arise from this range from `runtime not found org.gnome.Platform` to `Cannot access ~/Games` where the later is explicitilly set.

**Comment by hfiguiere** [comment]
Another example: https://twitter.com/zemarmot/status/1325015431481077760

tl;dr Gimp beta (2.99.2) on `flathub-beta` require installing the platform from `flathub` as a manual step. It seems that a lot of these users don't have any other flatpak app. (a workaround has been implemented, and it's not a good idea)

**Comment by doraskayo** [comment]
What would be an acceptable behavior to better handle this from Flatpak's perspective?

If we have further input from Flatpak project members, perhaps someone can pick it up and contribute a solution.

@alexlarsson, @mwleeds, @smcv, @TingPing

**Comment by nanonyme** [comment]
I think this possibly boils down to semantics. While necessary extensions here are provided by runtime, the extension point is declared in app manifest rather than runtime manifest. So they are essentially app extensions. They are still not hosted in app repo (which is beta repo) in this case but runtime repo (which is stable repo).
